src/Configuration/Config.AzureKeyVault/src/AzureKeyVaultConfigurationProvider.cs

```diff
@@ -7,77 +7,165 @@
+        public AzureKeyVaultConfigurationProvider(IKeyVaultClient client, string vault, IKeyVaultSecretManager manager, TimeSpan? reloadInterval = null)
         {
-            if (client == null)
+            _client = client ?? throw new ArgumentNullException(nameof(client));
+            _vault = vault ?? throw new ArgumentNullException(nameof(vault));
+            _manager = manager ?? throw new ArgumentNullException(nameof(manager));
+            if (_reloadInterval != null && _reloadInterval.Value == TimeSpan.Zero)
             {
-                throw new ArgumentNullException(nameof(client));
-            }
-            if (vault == null)
-            {
-                throw new ArgumentNullException(nameof(vault));
-            }
-            if (manager == null)
-            {
-                throw new ArgumentNullException(nameof(manager));
+                throw new ArgumentException(nameof(reloadInterval));
             }
 
-            _client = client;
-            _vault = vault;
-            _manager = manager;
+            _pollingTask = null;
+            _cancellationToken = new CancellationTokenSource();
+            _reloadInterval = reloadInterval;
         }
 
         public override void Load() => LoadAsync().ConfigureAwait(false).GetAwaiter().GetResult();
 
-        private async Task LoadAsync()
+        private async Task PollForSecretChangesAsync()
         {
-            var data = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            while (!_cancellationToken.IsCancellationRequested)
+            {
+                await WaitForReload();
+                await LoadAsync();
+            }
+        }
 
-            var secrets = await _client.GetSecretsAsync(_vault).ConfigureAwait(false);
-            var tasks = new List<Task<SecretBundle>>(secrets.Count());
+        protected virtual async Task WaitForReload()
+        {
+            await Task.Delay(_reloadInterval.Value.Milliseconds, _cancellationToken.Token);
+        }
+
+        private async Task LoadAsync()
+        {
+            var secretPage = await _client.GetSecretsAsync(_vault).ConfigureAwait(false);
+            var allSecrets = new List<SecretItem>(secretPage.Count());
             do
             {
-                tasks.Clear();
+                allSecrets.AddRange(secretPage.ToList());
+                secretPage = secretPage.NextPageLink != null ?
+                    await _client.GetSecretsNextAsync(secretPage.NextPageLink).ConfigureAwait(false) :
+                     null;
+            } while (secretPage != null);
 
-                foreach (var secretItem in secrets)
+            var tasks = new List<Task<SecretBundle>>();
+            var newLoadedSecrets = new Dictionary<string, LoadedSecret>();
+            var oldLoadedSecrets = Interlocked.Exchange(ref _loadedSecrets, null);
```

Comment:
Why Interlocked? It looks like _loadedSecrets is only used in this method and this method shouldn't be called concurrently (no?).

Reply:
In case someone calls `Reload` on configuration while automatic reload is in progress.

Reply:
oldLoadedSecrets being null will give you inconsistent results on the overlapping refreshes. What if we guarded this at a higher level? If someone calls reload while an automatic refresh is in progress, let them await the one that's already happening and use those results.

Reply:
> oldLoadedSecrets being null will give you inconsistent results on the overlapping refreshes

It is eventually consistent. Both reloads would retrieve the correct set of secrets.

> What if we guarded this at a higher level

I'm not sure if this option is simpler.